samples/event/chessboard.h

```diff
@@ -0,0 +1,90 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        samples/event/chessboard.h
+// Purpose:     Part of wxWidgets event sample, shows how to create a new wxEvent class
+// Author:      PB <pbfordev@gmail.com>
+// Created:     2019-10-29
+// Copyright:   (c) 2019 wxWidgets development team
+// Licence:     wxWindows licence
+/////////////////////////////////////////////////////////////////////////////
+
+#ifndef _WX_CHESSBOARD_H_
+
+#include "wx/wx.h"
+
+// A chessboard consists of 8x8 squares organized
+// in columns (called files) and rows (called ranks).
+// In a 2D view, files are labeled from left to right
+// 'a' to 'h' while ranks are numbered from bottom to
+// top 1 to 8.
+// ChessBoardEvent tells us something happened on
+// the chessboard. Either a square was clicked
+// (EVT_CHESSBOARD_CLICKED) or the mouse cursor was dragged
+// from one square to another (EVT_CHESSBOARD_DRAGGED).
+class ChessBoardEvent : public wxCommandEvent
+{
+public:
+    ChessBoardEvent(wxEventType commandEventType = wxEVT_NULL, int id = 0)
+        : wxCommandEvent(commandEventType, id),
+          m_file('a'), m_rank(1), m_fileTo('a'), m_rankTo(1)
+    {}
+
+    // EVT_CHESSBOARD_CLICKED: Returns File/Rank for the square clicked.
+    // EVT_CHESSBOARD_DRAGGED: Returns File/Rank for the square
+    //                         the mouse cursor was dragged from.
+    char GetFile() const    { return m_file; }
+    wxUint8 GetRank() const { return m_rank; }
+
+    // Valid only for EVT_CHESSBOARD_DRAGGED,
+    // returns File/Rank for the square where the mouse cursor was dragged to.
+    char GetFileTo() const    { return m_fileTo; }
+    wxUint8 GetRankTo() const { return m_rankTo; }
+
+    void SetFile(char file)    { m_file = file; }
+    void SetRank(wxUint8 rank) { m_rank = rank; }
+
+    void SetFileTo(char file)    { m_fileTo = file; }
+    void SetRankTo(wxUint8 rank) { m_rankTo = rank; }
+
+    virtual wxEvent* Clone() const wxOVERRIDE { return new ChessBoardEvent(*this); }
+private:
+    char  m_file;
+    wxUint8 m_rank;
+    char m_fileTo;
+    wxUint8 m_rankTo;
+};
+
+
+// Declare new events,
+// the matching definitions are in chessboard.cpp.
+wxDECLARE_EVENT(EVT_CHESSBOARD_CLICKED, ChessBoardEvent);
+wxDECLARE_EVENT(EVT_CHESSBOARD_DRAGGED, ChessBoardEvent);
+
+
+// @TODO: what to write here, the info in the docs seem to be incorrect?
+typedef void (wxEvtHandler::* ChessBoardEventFunction)(ChessBoardEvent&);
+#define ChessBoardEventHandler(func) wxEVENT_HANDLER_CAST(ChessBoardEventFunction, func)
+
```

Comment:
I wonder what the comment here should be. The code in docs reads

> // if you want to support old compilers you need to use some ugly macros:
> typedef void (wxEvtHandler::*MyPlotEventFunction)(MyPlotEvent&);
> #define MyPlotEventHandler(func) wxEVENT_HANDLER_CAST(MyPlotEventFunction, func)
> // if your code is only built using reasonably modern compilers, you could just
> // do this instead:
> #define MyPlotEventHandler(func) (&func)

but vadz told me that the part about old compilers is not true.

Reply:
Sorry, looking at this closer it seems that `wxEVENT_HANDLER_CAST` is needed, i.e. I think that if you remove it, the code won't compile even with recent gcc.

Reply:
But what should we actually write there in comments? Also, if I understand the macros correctly, these are needed only for event tables and `Connect()` and not when only` Bind()` is used?

Reply:
Yes, exactly. So the comment should say something like this:
```
// If the new event class is to be used with event table macros or the legacy Connect(),
// this macro must be used to cast the type of a function handling it to the type
// expected by the event table machinery, see its use in EVT_XXX below.
//
// Note that you can skip defining this macro entirely if you only plan to use Bind().
```